test/metabase/lib/datascript_test.cljc

```diff
@@ -0,0 +1,1100 @@
+                       {:db/id              -5
+                        :mbql.stage/source  -4}]
+   :mbql.query/stage0 -1
+   :mbql.query/stage  -5})
+
+(deftest ^:parallel stage-rules-test
+  (let [conn      (ld/fresh-conn)
+        tx        (d/transact! conn [sample-5-stages])
+        tempids   (:tempids tx)
+        query-eid (get tempids -10)
+        db        (:db-after tx)
+        datalog   '[:find ?stage .
+                    :in $ % ?query ?stage-number :where
+                    (stage-number ?query ?stage-number ?stage)]]
+    (is (= (get tempids -1)
+           (d/q datalog db ld/rules-stages query-eid 0)))
+    (is (= (get tempids -2)
+           (d/q datalog db ld/rules-stages query-eid 1)))
+    (is (= (get tempids -3)
+           (d/q datalog db ld/rules-stages query-eid 2)))
+    (is (= (get tempids -4)
+           (d/q datalog db ld/rules-stages query-eid 3)))
+    (is (= (get tempids -5)
+           (d/q datalog db ld/rules-stages query-eid 4)))
+    (is (= (get tempids -5)
+           (d/q datalog db ld/rules-stages query-eid -1)))))
+
+(deftest ^:parallel expr-test
+  (testing "[= entid literal]"
+    (is (=? {:mbql.clause/operator :=
+             :mbql.clause/argument [{:mbql/series 0, :mbql/ref [:metadata.field/id 12]}
+                                    {:mbql/series 1, :mbql/lit 7}]}
+            (ld/expr := [:metadata.field/id 12] 7))))
+  (testing "[= entid Entity]"
+    (is (=? {:mbql.clause/operator :=
+             :mbql.clause/argument [{:mbql/ref    [:metadata.field/id 12]
+                                     :mbql/series 0}
+                                    {:mbql/ref    8
+                                     :mbql/series 1}]}
+            (ld/expr := [:metadata.field/id 12] {:db/id 8, :metadata.column/source {:db/id 9}}))))
+  (testing "nesting and multiple arguments"
+    (testing "[= entid [= entid 6] 7 8 9]"
+      (is (=? {:mbql.clause/operator :=
+               :mbql.clause/argument [{:mbql/ref    [:metadata.field/id 12]
+                                       :mbql/series 0}
+                                      {:mbql/sub    {:mbql.clause/operator :+
+                                                     :mbql.clause/argument [{:mbql/ref    [:metadata.field/id 9001]
+                                                                             :mbql/series 0}
+                                                                            {:mbql/lit    6
+                                                                             :mbql/series 1}]}
+                                       :mbql/series 1}
+                                      {:mbql/lit 7, :mbql/series 2}
+                                      {:mbql/lit 8, :mbql/series 3}
+                                      {:mbql/lit 9, :mbql/series 4}]}
+              (ld/expr := [:metadata.field/id 12]
+                       (ld/expr :+ [:metadata.field/id 9001] 6)
+                       7 8 9))))))
```

Comment:
This test gives a good picture of how `:mbcl.clause/*` expression ASTs are stored - they're pretty clunky to write by hand, but `(ld/expr :+ lhs rhs)` below is just like `(lib/+ lhs rhs)` in MLv2.

(I didn't want to create a whole bunch of `ld/+` etc. functions that conflict with `clojure.core` in this PoC, but they can certainly be added.)